File: email-service/src/main/java/com/rj/ecommerce_email_service/messaging/listener/DeadLetterQueueListener.java
package com.rj.ecommerce_email_service.messaging.listener;

import lombok.extern.slf4j.Slf4j;
import org.springframework.amqp.core.Message;
import org.springframework.amqp.rabbit.annotation.RabbitListener;
import org.springframework.stereotype.Component;

@Component
@Slf4j
public class DeadLetterQueueListener {

    @RabbitListener(queues = RabbitMQDeadLetterConfig.EMAIL_DLQ)
    public void processFailedMessages(Message message) {
        log.error("Received dead letter message: {}", message);
        
        // Extract message headers to understand why it failed
        String xDeathHeader = message.getMessageProperties().getHeader("x-death") != null ? 
                message.getMessageProperties().getHeader("x-death").toString() : "No x-death header";
        
        log.error("Message failure reason: {}", xDeathHeader);
        
        // Here you could implement:
        // 1. Notification to admin
        // 2. Store failed messages in database for later analysis
        // 3. Attempt to fix and republish if possible
    }
}

File: payment-service/src/main/kotlin/com/rj/payment_service/listener/DeadLetterQueueListener.kt
package com.rj.payment_service.listener

import io.github.oshai.kotlinlogging.KotlinLogging
import org.springframework.amqp.core.Message
import org.springframework.amqp.rabbit.annotation.RabbitListener
import org.springframework.stereotype.Component

@Component
class DeadLetterQueueListener {

    companion object {
        private val logger = KotlinLogging.logger { }

    }

    @RabbitListener(queues = ["\${app.rabbitmq.dlq-queue}"])
    fun handleDeadLetterQueue(message: Message) {
        val originalQueue = message
            .messageProperties
            .headers["x-original-queue"].toString()

        val errorMessage = message
            .messageProperties
            .headers["x-original-error-message"].toString()
        // TODO:
        // Here you could:
        // 1. Send notifications to admin
        // 2. Store failed messages in database
        // 3. Implement custom recovery logic
        // 4. Forward to another service
    }
}

